### ISSUE TITLE
official support for uv

### ISSUE THREAD

**Issue by dbrtly** [direnv/direnv#1250]
**Is your feature request related to a problem? Please describe.**
I'd like direnv to layout python using astral-sh/uv in an elegant way.

**Describe the solution you'd like**

While this works, it breaks the prompt somehow. 
By default, without using direnv, uv uses the name of the base directory for the venv prompt.

``` --.config/direnv/direnvrc

realpath() {
    [[ $1 = /* ]] && echo "$1" || echo "$PWD/${1#./}"
}
layout_python-uv() {
    local python=${1:-python3}
    [[ $# -gt 0 ]] && shift
    unset PYTHONHOME
    if [[ -n $VIRTUAL_ENV ]]; then
        VIRTUAL_ENV=$(realpath "${VIRTUAL_ENV}")
    else
        local python_version
        python_version=$("$python" -c "import platform; print(platform.python_version())")
        if [[ -z $python_version ]]; then
            log_error "Could not detect Python version"
            return 1
        fi
        VIRTUAL_ENV=$PWD/.direnv/python-venv-$python_version
    fi
    export VIRTUAL_ENV
    if [[ ! -d $VIRTUAL_ENV ]]; then
        log_status "no venv found; creating $VIRTUAL_ENV"
        uv venv "$VIRTUAL_ENV" 
    fi

    PATH="${VIRTUAL_ENV}/bin:${PATH}"
    export PATH
    source "${VIRTUAL_ENV}/bin/activate"
}

```

**Describe alternatives you've considered**

**Additional context**

**Comment by inklesspen** [comment]
I've been using this in my `.config/direnv/lib/python_uv.sh` for a while now; it seems to work well:

```sh
layout_python_uv() {
  local python=${1:-python}
  [[ $# -gt 0 ]] && shift
  unset PYTHONHOME
  local python_version=$($python -V | cut -w -f 2 | cut -d . -f 1-2)
  if [[ -z $python_version ]]; then
      log_error "Could not find python's version"
      return 1
  fi

  if [[ -n "${VIRTUAL_ENV:-}" ]]; then
      local REPLY
      realpath.absolute "$VIRTUAL_ENV"
      VIRTUAL_ENV=$REPLY
  else
      VIRTUAL_ENV=$(direnv_layout_dir)/python-$python_version
  fi
  if [[ ! -d $VIRTUAL_ENV ]]; then
      uv venv -p $python "$@" "$VIRTUAL_ENV"
  fi
  export VIRTUAL_ENV
  PATH_add "$VIRTUAL_ENV/bin"
}
```

**Comment by pjz** [comment]
Couple of changes to run on linux:
```

layout_python_uv() {
  local python=${1:-python}
  [[ $# -gt 0 ]] && shift
  unset PYTHONHOME
  local python_path=$(uv python find $python)
  local python_version=$($python_path -V | cut -d' ' -f 2 | cut -d . -f 1-2)
  if [[ -z $python_version ]]; then
      log_error "Could not find python's version"
      return 1
  fi

  if [[ -n "${VIRTUAL_ENV:-}" ]]; then
      local REPLY
      realpath.absolute "$VIRTUAL_ENV"
      VIRTUAL_ENV=$REPLY
  else
      VIRTUAL_ENV=$(direnv_layout_dir)/python-$python_version
  fi
  export UV_PROJECT_ENVIRONMENT=$VIRTUAL_ENV
  if [[ ! -d $VIRTUAL_ENV ]]; then
      uv venv -p $python "$@" "$VIRTUAL_ENV"
  fi
  export VIRTUAL_ENV
  PATH_add "$VIRTUAL_ENV/bin"
}
```
1.  I dunno what version of `cut` supports `-w`, but mine doesn't.  Replaced with `-d' '`.
2. make a `python_path` tempvar so the version specified doesn't have to be on the $PATH as long as `uv python` knows where to find it.
3. export UV_PROJECT_ENVIRONMENT so that running `uv` alone with use the correct venv.

**Comment by zimbatm** [comment]
Would be good to add it to the [wiki](https://github.com/direnv/direnv/wiki).

If you believe the integration to be stable, sending a PR to add it to the stdlib would be fine as well.

**Comment by thernstig** [comment]
It has been added to the wiki by @doolio and @sephib and @consoull it seems.

(Though https://github.com/direnv/direnv/wiki was not updated to add it to the list there).

But would be great to add this to stdlib as I think it is quite obvious uv will take over more and more workflows.

**Comment by dpprdan** [comment]
I am wondering if we all share a common understanding of what an official support for uv should look like?

Here are some things that (IMHO) might be worth considering. These points were mostly triggered by looking at #1329, but it might just make sense to think about this more broadly and define some more concrete requirements for this issue?

1. `uv [init/add/sync/...]` vs. `uv [pip/venv]`

    The [uv docs state](https://docs.astral.sh/uv/getting-started/features/#the-pip-interface) that the uv pip interface is "intended to be used in legacy workflows or cases where the high-level commands do not provide enough control." For the direnv stdlib this boils down to the legacy workflows, I guess, because if you need more control you will probably need more control over direnv as well.

    It also follows that the [uv's project interface](https://docs.astral.sh/uv/getting-started/features/#projects) is the road ahead.

    For direnv this means that the project interface should the standard and the pip interface the special case, e.g. `layout uv` and `layout uv-pip`.

2. I wonder if it is really useful / necessary to build in `uv init` into direnv. `uv init` is only invoked once at the start of a project, so this only saves typing one command when initialising a project at the cost of adding complexity to direnv. Also, AFAICT project aren't initialised by direnv for any other language? Finally, it is fine if this is part of a template (i.e. https://github.com/direnv/direnv/wiki/Python), which every user can modify as they see fit, but building it into stdlib is a different cup of tea.

3. `uv` uses [Python version files](https://docs.astral.sh/uv/concepts/python-versions/#python-version-files) for specifying [which Python version to use](https://docs.astral.sh/uv/guides/projects/#python-version) when creating the project's virtual environment. Defining the python version in `.envrc` as well isn't DRY then (and the one in `.envrc` shouldn't be able to override the one in `.python-version` IMO). I think this is mostly an issue if one wants direnv to initialise the project (see 2.), so one more reason against that, I suppose?

4. Instead, it might be more useful to add [`uv sync`](https://docs.astral.sh/uv/reference/cli/#uv-sync) (and likewise `uv pip sync` for the `uv pip` workflow) to ensure that alle project dependencies are installed and up-to-date?

5. uv provides the [`uv run` command](https://docs.astral.sh/uv/reference/cli/#uv-run) to execute a command or script in the project environment. So with uv, activating a venv is [optional](https://docs.astral.sh/uv/reference/cli/#uv-venv). In fact the uv developers "[don't activate virtual environments anymore](https://x.com/charliermarsh/status/1837538829579796751) ([in favor of `uv run`](https://x.com/charliermarsh/status/1825958681042944065)) and at least [contemplate](https://github.com/astral-sh/uv/issues/1910#issuecomment-2256407301) that virtualenv activation won't find much use in a future Python ecosystem.

    I am sure that not everyone here will agree with the latter, but maybe virtualenv activation should be optional in direnv's uv workflow, e.g. a `use venv` function?

I guess what I am advocating for here is letting `uv` do the heavy lifting of managing projects, venvs and python versions and to not override it's defaults unless absolutely necessary. In fact IMHO a `uv sync` and a (for projects optional) `use/activate venv` [might just suffice](https://offby1.website/posts/uv-direnv-and-simple-envrc-files.html). Anyway, feel free to disagree!

**Comment by duckpuppy** [comment]
At least for the time being, activating the virtual environment also allows IDEs and other tools to see the dependencies included in the virtual environment. I use NeoVim with Ruff and without an activated venv ruff won't see the dependencies.

**Comment by dpprdan** [comment]
@duckpuppy I agree, that activating the virtual environment should be a feature, albeit possibly an optional one.

**Comment by inklesspen** [comment]
I still use and prefer the uv pip commands. I am uncomfortable with the idea of letting one tool (however rapid its development and adoption) take over the Python ecosystem.

Also there are a variety of situations (using meson-python to build packages, for example) where activating the virtualenv is necessary.

**Comment by inklesspen** [comment]
(see https://github.com/mesonbuild/meson-python/issues/630 for more on that)

**Comment by thernstig** [comment]
Activating the the virtual environment is to me the most important point that direnv should do. All the others are optional.

I don't think it should sync automatically. I do think it should install the Python version specified automatically if it is not installed, or rather warn about it. Please note that often the version can only be specified in `pyproject.toml`, not necessitating a `.python-version`. I'd consider using a `pyproject.toml` the more modern approach.

Ergo: It seems we have very different goals of what a default direnv implementation should do. To me activating the virtual environment seems the most reasonable one.

**Comment by inklesspen** [comment]
for the record i do not ever ask uv to download some random python binary. i use pyenv's python-build script (which i have installed as a standalone binary), so my .envrc usually looks something like this (with the function definition given [here](https://github.com/direnv/direnv/issues/1250#issuecomment-2253513786)):

```
layout python_uv /opt/snakes/python-3.12.7/bin/python
```

or of course sometimes more like this:
```
layout python_uv /opt/snakes/python-3.12.7/bin/python
load_prefix /opt/homebrew/opt/libressl
load_prefix /opt/homebrew/opt/postgresql@16
export POSTGRES_MAJOR_VERSION=`pg_config --version | perl -wnl -e '/\d+(?=\.)/ and print $&;'`

# Place the data directory inside the project directory
export PGDATA="$(pwd)/postgres-$POSTGRES_MAJOR_VERSION"
# Place Postgres' Unix socket inside the data directory
export PGHOST="$PGDATA"

if [[ ! -d "$PGDATA" ]]; then
	# If the data directory doesn't exist, create an empty one, and...
	initdb
	# ...configure it to listen only on the Unix socket, and...
	cat >> "$PGDATA/postgresql.conf" <<-EOF
		listen_addresses = ''
		unix_socket_directories = '$PGHOST'
	EOF
	# ...create a database using the name Postgres defaults to.
	echo "CREATE DATABASE $USER;" | postgres --single -E postgres
fi

export SQLALCHEMY_URL="postgresql://$USER:@/myproject"
export SQLALCHEMY_WARN_20=1
```

**Comment by offbyone** [comment]
While I would consider using a stdlib `uv` integration, in its absence I've [developed one of my own](https://offby1.website/posts/uv-direnv-and-simple-envrc-files.html) that suits me well.

If `uv` support is going in the direnv stdlib, I think it should be set up as much as possible to work _with_ the tool and not with a legacy function that is likely to be less well supported over time. Direnv is a stable tool, and the stdlib should look to the future, not the past.

`uv sync` is, even on first run, _ridiculously fast_; I have it run every time I chdir, to no performance cost. What I think it isn't, though, is a "layout"; there's nothing direnv should be managing about where its files go.

**Comment by dbrtly** [comment]
Uv has evolved significantly in 8 months. 

`uv sync` works for me. 

Fwiw, I use `uv run` by default without activating the virtual environment.

**Comment by duckpuppy** [comment]
`uv run` works as long as you're not using any code analysis tools like linters or LSP servers that need to have access to all the dependencies. For those right now we need to have an active venv.

**Comment by offbyone** [comment]
That's why #1352 also activates the venv.

**Comment by offbyone** [comment]
Can I get a sense from the direnv team what the decision criteria is on these PRs? I'd like to know how much more time I should throw at it, and who's gonna be the decider for if it gets merged.

**Comment by staticdev** [comment]
@zimbatm I see that there are problems with the wiki bash function when the project is already created and has a `pyproject.toml` file. Maybe would be great if we can have a built-in implementation like in https://github.com/direnv/direnv/pull/1352

I just wanted to ping on that since this topic here seems stopped ;)

**Comment by krupan** [comment]
Hi, I just sat down and got to know uv, and then read through most of the discussion here, and I just have to add my 2 cents.

First and foremost, direnv is a tool that sets environment variables when you cd into a directory, and restores them to what they were before when you cd out of that directory.  That's it.  Since the .envrc file that it uses is just a shell script, you can have direnv do some things other than that, but really that's beyond the scope of direnv.

We have had `layout python` available to put in our .envrc files for a while.  When you cd into a project directory, direnv:

1. creates a virtual environment if needed
2. sets environment variables equivalent to activating the virtual environment
 
Step 2 is basic direnv stuff, but step 1 is a little above and beyond normal direnv functionality.  It's a fast procedure and you know that's what you are getting when you use `layout python`

When you cd out of a project directory, direnv:

1. unsets/restores environment variables equivalent to deactivating the virtual environment

I too want to preserve this level of automation for my python projects but use uv instead venv/pip/etc.  What does that look like?  It's tricky because uv works differently than the usual python virtual environment flow.  The normal uv way of doing this is this:

Run `uv init project-name`, which creates a directory with a a toml file and a lock file for tracking project package dependencies, plus a whole bunch of other stuff like initializing a git repo, but it does not create a virtual environment.  The recommended uv way of proceeding is to just start running commands (like `python`), prefixing every command you run with `uv run`.  `uv run` basically:

1. creates a virtual environment if needed
2. activates the virtual environment
3. installs/removes packages to match what's in the toml
4. edits the lock file to record exactly what packages and version are installed
5. runs your command
6. deactivates the virtual environment and returns you to your shell prompt

If you are using uv this way, then direnv isn't necessary at all.  I personally do not like this style of working.  I have tools and scripts that don't know to prefix everything with `uv run`.  This is why direnv is my favorite tool in the world, it automatically sets up my shell environment so none of my tools or scripts need to know anything special, they just use `PATH` and whatever other environment variables they expect to be there.

To avoid using `uv run` you can run `uv sync` which does steps 1, 3, and 4.  And then you can activate/deactivate your virtual environment without using uv and do your work.  This fits the direnv with `layout python` workflow except that steps 3 and 4 are very much outside the scope of normal direnv functionality.  Step 3 especially can be very time consuming if it's downloading packages from the internet the first time.

So, you can drop down to the more manual and less recommended (by uv) command: `uv venv`.  This will remove a venv if it already exists and create a new one.  The removal is unfortunate, but because uv caches everything, if you run `uv sync` to install your packages again it will just do some fast file linking to reinstall the packages.  Or if you really don't like the removal you can 

So, in summary, I think to mimic `layout python` with uv, `layout uv` would only do something like this:

```
uv venv
source_env .venv/bin/activate
```

And then any other uv command you'd run on your own, if you choose.

**Comment by Mithrandir2k18** [comment]
So, I've had this probably dumb idea, I'm testing it right now to see if it can work for me, but it seems to work pretty well for the time being. I just  do:

```bash
uv run bash
```

<details>
  <summary>Just in case anybody is interested in how the `uv run bash` hack works, here it is.</summary>
This is a hack I'm testing, I provide no guarantees.

  Add this to your bashrc:
  
  ```bash
_auto_direnv_uv_shell() {
  if [[ -z "$USE_UV_RUN_BASH" && -n "$INSIDE_UV_RUN_BASH" ]]; then
    echo "Left project folder, exiting uv run bash"
    echo $(pwd) > "$UV_EXIT_CD_LOCATION"
    exit
  fi

  if [[ "$USE_UV_RUN_BASH" == "1" && -z "$INSIDE_UV_RUN_BASH" ]]; then
    if [[ "$UV_RUN_BASH_EXITED" == "1" ]]; then
      echo "Skipping re-entry: just exited uv shell"
      return
    fi

    export INSIDE_UV_RUN_BASH=1
    export UV_RUN_BASH_EXITED=0
    export UV_EXIT_CD_LOCATION="$XDG_RUNTIME_DIR/$(cat /proc/sys/kernel/random/uuid).txt"
    echo 'Entering `uv run bash` subshell...'

    uv run bash

    export UV_RUN_BASH_EXITED=1
    if [ -f $UV_EXIT_CD_LOCATION ]; then
        cd "$(head -1 $UV_EXIT_CD_LOCATION)"
        unset UV_RUN_BASH_EXITED
    fi
    unset INSIDE_UV_RUN_BASH
    unset UV_EXIT_CD_LOCATION
  fi
}

PROMPT_COMMAND="$PROMPT_COMMAND; _auto_direnv_uv_shell; "

  ```

Then to automatically enter a bash subshell with direnv, create a .envrc in a uv project with the following content:

```
export USE_UV_RUN_BASH=1
export UV_RUN_BASH_EXITED=0
```

Since this runs in subshells you'd use some bash history. For me it doesn't matter though as I'm using tmux and others are using https://github.com/atuinsh/atuin for that anyway.

This is not a recommended solution, there's probably other side-effects of it I'm not aware of yet and it is a bit messy. It's just a hack that's working for me right now.
</details>

**Comment by mschulkind** [comment]
I love the idea. I've had just a slightly nicer experience with:
```
[ -d .venv ] || uv venv
source_env .venv/bin/activate
```

**Comment by injust** [comment]
Folding https://github.com/direnv/direnv/issues/1477 into this issue, so just noting that the current https://github.com/direnv/direnv/wiki/Python#uv does not work with `global.strict_env`.

```
direnv: loading /path/to/.envrc
~/.config/direnv/direnvrc:7: VIRTUAL_ENV: unbound variable
```

**Comment by injust** [comment]
This is what I've been using for some time now:

```bash
uv sync
source_env .venv/bin/activate
watch_file pyproject.toml
```

`uv sync` is fast enough to run every time (no need to check if `.venv/` already exists, plus this keeps your venv in sync).

It also watches the venv activation script and pyproject.toml (substitute for requirements.txt or any other dependency file you use), keeping the venv updated if dependencies change or the venv is recreated.
